commands.go

```diff
@@ -130,67 +166,104 @@ func ctfCommandCallback(s *discordgo.Session, i *discordgo.InteractionCreate) {
 				},
 			},
 		})
+
+		s.ChannelMessageSendComplex(channel.ID, &discordgo.MessageSend{
+			Content:    ctfName,
+			Components: leaveRespComponent,
+		})
 	case "join":
-		guild, err := s.Guild(i.GuildID)
-		if err != nil {
-			s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
-				Type: discordgo.InteractionResponseChannelMessageWithSource,
-				Data: &discordgo.InteractionResponseData{
-					Flags:   1 << 6,
-					Content: "Couldn't find the calling Guild",
-				},
-			})
+		if len(data.Options) > 0 && len(data.Options[0].Options) > 0 {
+			ctfName := data.Options[0].Options[0].StringValue()
+			joinCTFCallback(s, i, ctfName)
+			// We respond in the function call, no need to attempt to respond twice.
 			return
 		}
-		ctfName := i.Message.Content
-		callingUser := i.Member.User
-		log.Printf("Adding user %s to CTF %s\n", callingUser.Username, ctfName)
-
-		// Find the role of the CTF and add the user to it
-		var ctfRole *discordgo.Role
-		for _, r := range guild.Roles {
-			if r.Name == ctfName {
-				ctfRole = r
-			}
-		}
-		// Check if the role was ever found and correct if not
-		if ctfRole == nil {
-			s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
-				Type: discordgo.InteractionResponseChannelMessageWithSource,
-				Data: &discordgo.InteractionResponseData{
-					Flags:   1 << 6,
-					Content: "Role " + ctfName + " does not exist. Try creating it, first!",
-				},
-			})
+
+	case "leave":
+		if len(data.Options) > 0 && len(data.Options[0].Options) > 0 {
+			ctfName := data.Options[0].Options[0].StringValue()
+			leaveCTFCallback(s, i, ctfName)
+			// We respond in the function call, no need to attempt to respond twice.
 			return
 		}
+	}
 
-		s.GuildMemberRoleAdd(guild.ID, callingUser.ID, ctfRole.ID)
+	// Send back the status reply
+	err := s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
+		Type: discordgo.InteractionResponseChannelMessageWithSource,
+		Data: &discordgo.InteractionResponseData{
+			Content:    respContent,
+			Components: respComponents,
+		},
+	})
+	if err != nil {
+		log.Print(err)
+	}
+}
+
+func joinCTF(s *discordgo.Session, i *discordgo.InteractionCreate) {
+
+	ctfName := i.Message.Content
+	joinCTFCallback(s, i, ctfName)
```

Comment:
There might be a cleaner way to do this without two chained functions. Thoughts?

Reply:
Resolved in discussion